service/worker/archiver/client_v2.go

```diff
@@ -0,0 +1,254 @@
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+//go:generate mockgen -copyright_file ../../../LICENSE -package $GOPACKAGE -source $GOFILE -destination client_v2_mock.go
+
+package archiver
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	commonpb "go.temporal.io/api/common/v1"
+	enumspb "go.temporal.io/api/enums/v1"
+	"go.uber.org/multierr"
+
+	archiverspb "go.temporal.io/server/api/archiver/v1"
+	carchiver "go.temporal.io/server/common/archiver"
+	"go.temporal.io/server/common/archiver/provider"
+	"go.temporal.io/server/common/dynamicconfig"
+	"go.temporal.io/server/common/log"
+	"go.temporal.io/server/common/log/tag"
+	"go.temporal.io/server/common/metrics"
+	"go.temporal.io/server/common/primitives/timestamp"
+	"go.temporal.io/server/common/quotas"
+	"go.temporal.io/server/common/sdk"
+	"go.temporal.io/server/common/searchattribute"
+)
+
+type (
+	Request struct {
+		ShardID     int32
+		NamespaceID string
+		Namespace   string
+		WorkflowID  string
+		RunID       string
+
+		// history archival
+		BranchToken          []byte
+		NextEventID          int64
+		CloseFailoverVersion int64
+		HistoryURI           string
+
+		// visibility archival
+		WorkflowTypeName string
+		StartTime        time.Time
+		ExecutionTime    time.Time
+		CloseTime        time.Time
+		Status           enumspb.WorkflowExecutionStatus
+		HistoryLength    int64
+		Memo             *commonpb.Memo
+		SearchAttributes *commonpb.SearchAttributes
+		VisibilityURI    string
+
+		// archival targets: history and/or visibility
+		Targets       []ArchivalTarget
+		CallerService string
+	}
+
+	Response struct {
+	}
+
+	// ClientV2 is used to archive workflow history and visibility data. If a target fails, it returns an error, unlike
+	// Client, which will try to signal an archival workflow whenever an error occurrs.
+	ClientV2 interface {
+		Archive(context.Context, *Request) (*Response, error)
+	}
+
+	clientV2 struct {
+		metricsScope     metrics.Scope
+		logger           log.Logger
+		sdkClientFactory sdk.ClientFactory
+		numWorkflows     dynamicconfig.IntPropertyFn
+		signalTimeout    dynamicconfig.DurationPropertyFn
+		rateLimiter      quotas.RateLimiter
+		archiverProvider provider.ArchiverProvider
+	}
+)
+
+// NewClientV2 creates a new ClientV2
+func NewClientV2(
```

Comment:
should the client locate in the history service? If so, it should remove the v2  in the name.

Reply:
Moved into history/archival